torchtnt/framework/callbacks/torchsnapshot_saver.py

```diff
@@ -111,12 +109,41 @@ def on_train_epoch_end(self, state: State, unit: TTrainUnit) -> None:
         # TODO: discuss whether this path should be customized
         global_step = train_progress.num_steps_completed
         snapshot_path = _get_snapshot_save_path(self._dirpath, epoch, global_step)
-
-        # TODO: use async take and ensure pending snapshot is done before saving the next one
-        torchsnapshot.Snapshot.take(
-            snapshot_path, app_state=app_state, replicated=list(self._replicated)
+        self.async_snapshot(snapshot_path, app_state, wait=True)
+
+    def on_train_end(self, state: State, unit: TTrainUnit) -> None:
+        self.wait()
+
+    def on_exception(self, state: State, unit: TTrainUnit, exc: BaseException) -> None:
+        self.wait()
+
+    def wait(self) -> None:
+        if self._pending is not None:
+            self._pending.wait()
+
+    def async_snapshot(
+        self, snapshot_path: str, app_state: Dict[str, _TStateful], *, wait: bool
+    ) -> bool:
```

Comment:
mark as protected since we dont expect callers to directly call these methods
```suggestion
    def _wait(self) -> None:
        if self._pending is not None:
            self._pending.wait()

    def _async_snapshot(
        self, snapshot_path: str, app_state: Dict[str, _TStateful], *, wait: bool
    ) -> bool:
```

Reply:
done, but note I'll totally call those from my user code, because I want to override the trigger `on_train_step_end` but not they snapshotting itself.

Reply:
could you describe the motivation for overriding the trigger? is it something we can make easier through this class itself?

Reply:
First idea is that I want to change when to checkpoint: every x minutes instead of every x steps. I think this provide a more intuitive way of setting this parameter. Not sure if researchers will like this one.
Second is that I want to checkpoint after each validation and save the metrics with the model, and keep a symlink to the "best" checkpoint.